zapcore/syslog_encoder.go

```diff
@@ -0,0 +1,274 @@
+package zapcore
+
+import (
+	"log/syslog"
+	"os"
+	"sync"
+	"time"
+
+	"go.uber.org/zap/buffer"
+	"go.uber.org/zap/internal/bufferpool"
+)
+
+// /usr/include/sys/syslog.h
+const (
+	severityMask   = 0x07
+	facilityMask   = 0xf8
+	syslogNilValue = "-"
+	ceeCookie      = "@cee:"
+)
+
+var _ Encoder = &syslogEncoder{}
+
+var _syslogPool = sync.Pool{New: func() interface{} {
+	return &syslogEncoder{}
+}}
+
+type SyslogEncoderConfig struct {
+	// CEECookie assists syslog parsers to know that the msg is structured data. By default, don't set the cee cookie.
+	CEECookie bool
+	Facility  syslog.Priority
+	Hostname  string
+	PID       int
+	Tag       string
+}
+
+type syslogEncoder struct {
+	cee      bool
+	facility syslog.Priority
+	hostname string // TODO make configurable
+	je       *jsonEncoder
+	pid      int
+	tag      string // TODO make configurable
+}
+
+func getSyslogEncoder() *syslogEncoder {
+	return _syslogPool.Get().(*syslogEncoder)
+}
+
+func putSyslogEncoder(enc *syslogEncoder) {
+	enc.cee = false
+	enc.facility = 0
+	enc.hostname = ""
+	enc.je = nil
+	enc.pid = 0
+	enc.tag = ""
+	_jsonPool.Put(enc)
```

Comment:
If we can embed, then we just need to reset all embedded fields here.

Consider doing `*enc = syslogEncoder{}` rather than needing to list all the fields.